src/graph/SetExecutor.cpp

```diff
@@ -0,0 +1,392 @@
+    right_ = makeTraverseExecutor(sentence_->right());
+    DCHECK(left_ != nullptr);
+    DCHECK(right_ != nullptr);
+
+    setLeft();
+    setRight();
+
+    status = left_->prepare();
+    if (!status.ok()) {
+        FLOG_ERROR("Prepare executor `%s' failed: %s",
+                    left_->name(), status.toString().c_str());
+        return status;
+    }
+
+    status = right_->prepare();
+    if (!status.ok()) {
+        FLOG_ERROR("Prepare executor `%s' failed: %s",
+                    right_->name(), status.toString().c_str());
+        return status;
+    }
+
+    return Status::OK();
+}
+
+void SetExecutor::setLeft() {
+    auto onFinish = [] () {
+        return;
+    };
+
+    futures_.emplace_back(leftP_.getFuture());
+    auto onResult = [this] (std::unique_ptr<InterimResult> result) {
+    this->leftResult_ = std::move(result);
+        VLOG(3) << "Left result set.";
+        leftP_.setValue();
+    };
+
+    auto onError = [this] (Status s) {
+        VLOG(3) << "Left error:" << s.toString();
+        leftS_ = std::move(s);
+        leftP_.setValue();
+    };
+
+    left_->setOnResult(onResult);
+    left_->setOnFinish(onFinish);
+    left_->setOnError(onError);
+}
+
+void SetExecutor::setRight() {
+    auto onFinish = [] () {
+        return;
+    };
+
+    futures_.emplace_back(rightP_.getFuture());
+    auto onResult = [this] (std::unique_ptr<InterimResult> result) {
+        this->rightResult_ = std::move(result);
+        VLOG(3) << "Right result set.";
+        rightP_.setValue();
+    };
+
+    auto onError = [this] (Status s) {
+        VLOG(3) << "Right error: " << s.toString();
+        rightS_ = std::move(s);
+        rightP_.setValue();
+    };
+
+    right_->setOnResult(onResult);
+    right_->setOnFinish(onFinish);
+    right_->setOnError(onError);
+}
+
+void SetExecutor::execute() {
+    auto *runner = ectx()->rctx()->runner();
+    runner->add([this] () mutable { left_->execute(); });
+    runner->add([this] () mutable { right_->execute(); });
+
+    auto cb = [this] (auto &&result) {
+        UNUSED(result);
+        if (!leftS_.ok() || !rightS_.ok()) {
+            std::string msg;
+            msg += "left: ";
+            msg += leftS_.toString();
+            msg += " right: ";
+            msg += rightS_.toString();
+            onError_(Status::Error(msg));
+            return;
+        }
+
+        if (leftResult_ == nullptr && rightResult_ == nullptr) {
+            VLOG(3) << "Set op no input.";
+            onEmptyInputs();
+            return;
+        }
+
+        // Set op share the same priority, they would execute sequentially.
+        switch (sentence_->op()) {
+            case SetSentence::Operator::UNION:
+                doUnion();
+                break;
+            case SetSentence::Operator::INTERSECT:
+                doIntersect();
+                break;
+            case SetSentence::Operator::MINUS:
+                doMinus();
+                break;
+            default:
```

Comment:
missing exception handling, do  onError() here

Reply:
It should definitely not reach here, I will add fatal log here.